lib/TokenManager.js

```diff
@@ -120,25 +129,31 @@ function refresh(sdk, tokenMgmtRef, storage, key) {
   }
 
   // Remove existing autoRefresh timeout for this key
-  clearRefreshTimeout(tokenMgmtRef, key);
-
-  return sdk.token.refresh(token)
-  .then(function(freshToken) {
-    add(sdk, tokenMgmtRef, storage, key, freshToken);
-    tokenMgmtRef.emitter.emit('refreshed', key, freshToken, token);
-    return freshToken;
-  })
-  .fail(function(err) {
-    if (err.name === 'OAuthError') {
-      remove(tokenMgmtRef, storage, key);
-      emitExpired(tokenMgmtRef, key, token);
-    }
-    throw err;
-  });
+  clearExpireEventTimeout(tokenMgmtRef, key);
+
+  // Store the refresh promise state, to avoid refreshing again
+  if (!tokenMgmtRef.refreshPromise[key]) {
+    tokenMgmtRef.refreshPromise[key] = sdk.token.refresh(token)
+    .then(function(freshToken) {
+      add(tokenMgmtRef, storage, key, freshToken);
```

Comment:
Is there a reason we're deleting `refreshPromise[key]` in `add` and not here? Add can be called by flows outside of the refresh scenario.